components/ntp/manpages/ntp.conf.4
'\" te
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright (c) 2009, 2011, Oracle and/or its affiliates. All rights reserved.
.\"
.TH "ntp.conf" "4" "" "" "File Formats"
.SH NAME
ntp.conf \- Configuration file for the NTP Daemon.
.SH DESCRIPTION
.LP
The \fB\&ntp.conf\fR file contains the directives used by the \fBntpd\fR to
configure itself. 
.SS "Configuration Commands"
.sp 1.5
.ne 2
.mk
.na
\fBserver \fIaddress\fP \fR[options ...]\fP\fR
.br
\fBpeer \fIaddress\fP \fR[options ...]\fP\fR
.br
\fBbroadcast \fIaddress\fP \fR[options ...]\fP\fR
.br
\fBmanycastclient \fIaddress\fP \fR[options ...]\fP\fR
.br
\fR\fBpool \fIaddress\fP \fR[options ...]\fP\fR
.ad
.sp .25
.RS 4n
These commands specify the time server name or address to be used and the mode in which to operate. The \fIaddress\fP can be either a DNS name or a IPv4 or IPv6 address in standard notation. In general, multiple commands of each type can be used for different server and peer addresses or multicast groups.
.RE
.sp .3
.ne 2
.mk
.na
\fBserver\fR
.ad
.sp .25
.RS 4n
For type s and r addresses (only), this command mobilizes a persistent client mode association with the specified remote server or local reference clock. If the \fBpreempt\fR flag is specified, a preemptable client mode association is mobilized instead.
.RE
.sp .3
.ne 2
.mk
.na
\fBpeer\fR
.ad
.sp .25
.RS 4n
For type s addresses (only), this command mobilizes a persistent symmetric-active mode association with the specified remote peer.
.RE
.sp .3
.ne 2
.mk
.na
\fBbroadcast\fR
.ad
.sp .25
.RS 4n
For type b and m ddresses (only), this command mobilizes a persistent broadcast or multicast server mode association. Note that type b messages go only to the interface specified, but type m messages go to all interfaces.
.RE
.sp .3
.ne 2
.mk
.na
\fBmanycastclient\fR
.ad
.sp .25
.RS 4n
For type m addresses (only), this command mobilizes a manycast client mode association for the multicast group address specified. In this mode the address must match the address specified on the \fBmanycastserver\fR command of one or more designated manycast servers.
.RE
.sp .3
.ne 2
.mk
.na
\fBpool\fR
.ad
.sp .25
.RS 4n
For type s messages (only) this command mobilizes a client mode association for servers implementing the pool automatic server discovery scheme described on the Association Management page at file:///usr/share/doc/ntp/assoc.html. The address is a DNS name in the form \fB\fIarea\fP.pool.ntp.org\fR, where \fB\fIarea\fP\fR is a qualifier designating the server geographic area such as \fBus\fR or \fBeurope\fR.
.SS "Command Options"
Each of the above configuation commands takes zero or more options from the list below:
.sp .3
.ne 2
.mk
.na
\fBautokey\fR
.ad
.sp .25
.RS 4n
Send and receive packets authenticated by the autokey scheme described in the Authentication Options page at file:///usr/share/doc/ntp/authopt.html. This option is valid only with \fBserver\fR and \fBpeer\fR commands and type s addresses. It is incompatible with the \fBkey\fR option.
.RE
.sp .3
.ne 2
.mk
.na
\fBburst\fR
.ad
.sp .25
.RS 4n
When the server is reachable, send a burst of six packets instead of the usual one. The packet spacing is normally 2 s; however, the spacing between the first and second packets can be changed with the fBcalldelay\fR command to allow additional time for a modem or ISDN call to complete. This option is valid only with only the \fBserver\fR command and type s addressesa. It is a recommended option when the \fBmaxpoll\fR option is greater than 10 (1024 s).
.RE
.sp .3
.ne 2
.mk
.na
\fBiburst\fR
.ad
.sp .25
.RS 4n
When the server is unreachable, send a burst of eight packets instead of the usual one. The packet spacing is normally 2 s; however, the spacing between the first and second packets can be changed with the \fBcalldelay\fR command to allow additional time for a modem or ISDN call to complete. This option is valid only with the \fBserver\fR command and type s addresses. It is a recommended option with this command.
.RE
.sp .3
.ne 2
.mk
.na
\fBkey\fR \fIkey\fP\fR
.ad
.sp .25
.RS 4n
Send and receive packets authenticated by the symmetric key scheme described in the Authentication Options page at file:///usr/share/doc/ntp/authopt.html. This option is valid only with \fBserver\fR and \fBpeer\fR commands and type s addresses. The \fIkey\fP specifies the key identifier with values from 1 to 65534, inclusive. This option is incompatible with the \fBautokey\fR option.
.RE
.sp .3
.ne 2
.mk
.na
\fBminpoll \fIminpoll
.br
\fP\fR\fBmaxpoll \fImaxpoll\fP\fR
.ad
.sp .25
.RS 4n
These options specify the minimum and maximum poll intervals for NTP messages, in seconds as a power of two. The maximum poll interval defaults to 10 (1024 s), but can be increased by the \fBmaxpoll\fR option to an upper limit of 17 (36 h). The minimum poll interval defaults to 6 (64 s), but can be decreased by the \fBminpoll\fR option to a lower limit of 4 (16 s). These option are valid only with the \fBserver\fR and \fBpeer\fR commands and type s addresses.
.RE
.sp .3
.ne 2
.mk
.na
\fBmode \fIoption\fP\fR
.ad
.sp .25
.RS 4n
Pass the \fB\fIoption\fP\fR to a reference clock driver, where \fB\fIoption\fP\fR is an integer in the range from 0 to 255, inclusive. This option is valid only with the \fBserver\fR command and type r addresses.
.RE
.sp .3
.ne 2
.mk
.na
\fBnoselect\fR
.ad
.sp .25
.RS 4n
Marks the server or peer to be ignored by the selection algorithm but visible to the monitoring program. This option is ignored with the \fBbroadcast\fR command.
.RE
.sp .3
.ne 2
.mk
.na
\fBpreempt\fR
.ad
.sp .25
.RS 4n
Specifies the association as preemptable rather than the default persistent. This option is ignored with the  \fBbroadcast\fR command and is most useful with the \fBmanycastclient\fR and \fBpool\fR commands.
.RE
.sp .3
.ne 2
.mk
.na
\fBprefer\fR
.ad
.sp .25
.RS 4n
Mark the server as preferred. All other things being equal, this host will be chosen for synchronization among a set of correctly operating hosts. See the Mitigation Rules page at file:///usr/share/doc/ntp/prefer.html for further information. This option is valid only with the \fBserver\fR and \fBpeer\fR commands.
.RE
.sp .3
.ne 2
.mk
.na
\fBtrue\fR
.ad
.sp .25
.RS 4n
Mark the association to assume truechimer status; that is, always survive the selection and clustering algorithms. This option can be used with any association, but is most useful for reference clocks with large jitter on the serial port and precision pulse-per-second (PPS) signals. Caution: this option defeats the algorithms designed to cast out falsetickers and can allow these sources to set the system clock. This option is valid only with the \fBserver\fR and \fBpeer\fR commands.
.RE
.sp .3
.ne 2
.mk
.na
\fBttl \fIttl\fP\fR
.ad
.sp .25
.RS 4n
This option specifies the time-to-live \fIttl\fP for the \fBbroadcast\fR commmand and the maximum \fIttl\fP for the expanding ring search used by the \fBmanycastclient\fR command. Selection of the proper value, which defaults to 127, is something of a black art and should be coordinated with the network administrator.
.RE
.sp .3
.ne 2
.mk
.na
\fBversion \fIversion\fP\fR
.ad
.sp .25
.RS 4n
Specifies the version number to be used for outgoing NTP packets. Versions 1-4 are the choices, with version 4 the default.
.RE
.SS "Auxilliary Commands"
.sp .3
.ne 2
.mk
.na
\fBbroadcastclient [novolley]\fR
.ad
.sp .25
.RS 4n
Enable reception of broadcast server messages to any local interface (type b address). Ordinarily, upon receiving a message for the first time, the broadcast client measures the nominal server propagation delay using a brief client/server exchange, after which it continues in listen-only mode. If the \fBnovolley\fR keyword is present, the exchange is not used and the value specified in the \fBbroadcastdelay\fR command is used or, if the \fBbroadcastdelay\fR command is not used, the default 4.0 ms. Note that, in order to avoid accidental or malicious disruption in this mode, both the server and client should operate using symmetric key or public key authentication as described in the Authentication Options page at file:///usr/share/doc/ntp/authopt.html. Note that the \fBnovolley\fR keyword is incompatible with public key authentication.
.RE
.sp .3
.ne 2
.mk
.na
\fBmanycastserver \fIaddress\fP [...]\fR
.ad
.sp .25
.RS 4n
Enable reception of manycast client messages (type m)to the multicast group address(es) (type m) specified. At least one address is required. Note that, in order to avoid accidental or malicious disruption, both the server and client should operate using symmetric key or public key authentication as described in the Authentication Options page at file:///usr/share/doc/ntp/authopt.html.
.RE
.sp .3
.ne 2
.mk
.na
\fBmulticastclient \fIaddress\fP [...]\fR
.ad
.sp .25
.RS 4n
Enable reception of multicast server messages to the multicast group address(es) (type m) specified. Upon receiving a message for the first time, the multicast client measures the nominal server propagation delay using a brief client/server exchange with the server, then enters the broadcast client mode, in which it synchronizes to succeeding multicast messages. Note that, in order to avoid accidental or malicious disruption in this mode, both the server and client should operate using symmetric key or public key authentication as described in the Authentication Options page at file:///usr/share/doc/ntp/authopt.html.
.RE
.SS "Reference Clock Commands"
.sp .3
.ne 2
.mk
.na
\fBserver 127.127.\fIt.u\fP [prefer] [mode \fIint\fP] [minpoll \fIint\fP] [maxpoll \fIint\fP]\fR
.ad
.sp .25
.RS 4n
This command can be used to configure reference clocks in special ways. The options are interpreted as follows:
.RE
.sp .3
.ne 2
.mk
.na
\fBprefer\fR
.ad
.sp .25
.RS 4n
Marks the reference clock as preferred. All other things being equal, this host will be chosen for synchronization among a set of correctly operating hosts. See the Mitigation Rules page at file:///usr/share/doc/ntp/prefer.html for further information.
.RE
.sp .3
.ne 2
.mk
.na
\fBmode \fIint\fP\fR
.ad
.sp .25
.RS 4n
Specifies a mode number which is interpreted in a device-specific fashion. For instance, it selects a dialing protocol in the ACTS driver and a device subtype in the \fBparse\fR drivers.
.RE
.sp .3
.ne 2
.mk
.na
\fBminpoll \fIint\fP\fR
.RE
.sp .3
.ne 2
.mk
.na
\fBmaxpoll \fIint\fP\fR
.ad
.sp .25
.RS 4n
These options specify the minimum and maximum polling interval for reference clock messages in seconds, interpreted as dual logarithms (2 ^ x). For most directly connected reference clocks, both \fBminpoll\fR and \fBmaxpoll\fR default to 6 (2^16 = 64 s). For modem reference clocks, \fBminpoll\fR defaults to 10 (2^10 = 1024 s = 17.1 m) and \fBmaxpoll\fR defaults to 14 (2^14 = 16384 s = 4.25 h). The allowable range is 4 (16 s) to 17 (36.4 h) inclusive.
.RE
.sp .3
.ne 2
.mk
.na
\fBfudge 127.127.\fIt.u\fP [time1 \fIsec\fP] [time2 \fIsec\fP] [stratum \fIint\fP] [refid \fIstring\fP] [mode \fIint\fP] [flag1 0|1] [flag2 0|1] [flag3 0|1] [flag4 0|1]\fR
.ad
.sp .25
.RS 4n
This command can be used to configure reference clocks in special ways. It must immediately follow the \fBserver\fR command which configures the driver. Note that the same capability is possible at run time using the \fBntpdc\fR program. The options are interpreted as follows:
.RE
.sp .3
.ne 2
.mk
.na
\fBtime1 \fIsec\fP\fR
.ad
.sp .25
.RS 4n
Specifies a constant to be added to the time offset produced by the driver, a fixed-point decimal number in seconds. This is used as a calibration constant to adjust the nominal time offset of a particular clock to agree with an external standard, such as a precision PPS signal. It also provides a way to correct a systematic error or bias due to serial port or operating system latencies, different cable lengths or receiver internal delay. The specified offset is in addition to the propagation delay provided by other means, such as internal DIPswitches. Where a calibration for an individual system and driver is available, an approximate correction is noted in the driver documentation pages.
.LP
Note: in order to facilitate calibration when more than one radio clock or PPS signal is supported, a special calibration feature is available. It takes the form of an argument to the \fBenable\fR command and operates as described in the Reference Clock Drivers page at file:///usr/share/doc/ntp/refclock.html.
.RE
.sp .3
.ne 2
.mk
.na
\fBtime2 \fIsecs\fP\fR
.ad
.sp .25
.RS 4n
Specifies a fixed-point decimal number in seconds, which is interpreted in a driver-dependent way. See the descriptions of specific drivers in the Reference Clock Drivers page at file:///usr/share/doc/ntp/refclock.html.
.RE
.sp .3
.ne 2
.mk
.na
\fBstratum \fIint\fP\fR
.ad
.sp .25
.RS 4n
Specifies the stratum number assigned to the driver, an integer between 0 and 15. This number overrides the default stratum number ordinarily assigned by the driver itself, usually zero.
.RE
.sp .3
.ne 2
.mk
.na
\fBrefid \fIstring\fP\fR
.ad
.sp .25
.RS 4n
Specifies an ASCII string of from one to four characters which defines the reference identifier used by the driver. This string overrides the default identifier ordinarily assigned by the driver itself.
.RE
.sp .3
.ne 2
.mk
.na
\fBmode \fIint\fP\fR
.ad
.sp .25
.RS 4n
Specifies a mode number which is interpreted in a device-specific fashion. For instance, it selects a dialing protocol in the ACTS driver and a device subtype in the \fBparse\fR drivers.
.RE
.sp .3
.ne 2
.mk
.na
\fBflag1 flag2 flag3 flag4\fR
.ad
.sp .25
.RS 4n
These four flags are used for customizing the clock driver. The interpretation of these values, and whether they are used at all, is a function of the particular clock driver. However, by convention \fBflag4\fR is used to enable recording monitoring data to the \fBclockstats\fR file configured with the \fBfilegen\fR command. 
.RE
.SS "Authentication Commands"
.sp .3
.ne 2
.mk
.na
\fBautokey [\fIlogsec\fP]\fR
.ad
.sp .25
.RS 4n
Specifies the interval between regenerations of the session key list used with the Autokey protocol. Note that the size of the key list for each association depends on this interval and the current poll interval. The default value is 12 (4096 s or about 1.1 hours). For poll intervals above the specified interval, a session key list with a single entry will be regenerated for every message sent.
.RE
.sp .3
.ne 2
.mk
.na
\fBcontrolkey \fIkey\fP\fR
.ad
.sp .25
.RS 4n
Specifies the key identifier to use with the \fBntpq\fR utility, which uses the standard protocol defined in RFC-1305. The \fB\fIkey\fP\fR argument is the key identifier for a trusted key, where the value can be in the range 1 to 65,534, inclusive.
.RE
.sp .3
.ne 2
.mk
.na
\fBcrypto [randfile \fIfile\fP] [host \fIname\fP] [ident \fIname\fP] [pw \fIpassword\fP]\fR
.ad
.sp .25
.RS 4n
This command requires the OpenSSL library. It activates public key cryptography and loads the required public/private encryption and sign kyes and public certificat. If one or more files are left unspecified, the default names are used as described below. Unless the complete path and name of the file are specified, the location of a file is relative to the keys directory specified in the \fBkeysdir\fR command or default \fB/etc/inet\fR. Following are the subcommands.
.RE
.sp .3
.ne 2
.mk
.na
\fBhost \fIname\fP\fR
.ad
.sp .25
.RS 4n
Specifies the host name used in the host key link \fBntpkey_host_\fIname\fP\fR, sign key link \fBntpkey_sign_\fIname\fP\fR and certificate link \fBntpkey_cert_\fIname\fP\fR. The \fBntp-keygen\fR program automatically installs these links to the most recently generated files.
.RE
.sp .3
.ne 2
.mk
.na
\fBident \fIname\fP\fR
.ad
.sp .25
.RS 4n
Specifies the group name used in the identity key link \fBntpkey_\fIkey\fP_\fIname\fP\fR, where \fB\fIkey\fP\fR identifies the key type described on the \fBntp-keygen\fR page. The \fBntp-keygen\fR program automatically installs these links to the most recently generated files.
.RE
.sp .3
.ne 2
.mk
.na
\fBpw \fIpassword\fP\fR
.ad
.sp .25
.RS 4n
Specifies the password to decrypt files previously encrypted by the \fBntp-keygen\fR program.
.RE
.sp .3
.ne 2
.mk
.na
\fBrandfile \fIfile\fP\fR
.ad
.sp .25
.RS 4n
Specifies the location of the random seed file used by the OpenSSL library. The defaults are described on the \fBntp-keygen(1m)\fR man page.
.RE
.sp .3
.ne 2
.mk
.na
\fBkeys \fIkeyfile\fP\fR
.ad
.sp .25
.RS 4n
Specifies the complete path to the MD5 key file containing the keys and key identifiers used by \fBntpd\fR, \fBntpq\fR and \fBntpdc\fR when operating with symmetric key cryptography. This is the same operation as the \fB-k \fRcommand line option.
.RE
.sp .3
.ne 2
.mk
.na
\fBkeysdir \fIpath\fP\fR
.ad
.sp .25
.RS 4n
This command specifies the default directory path for cryptographic keys, parameters and certificates. The default is \fB/etc/inet/\fR.
.RE
.sp .3
.ne 2
.mk
.na
\fBrequestkey \fIkey\fP\fR
.ad
.sp .25
.RS 4n
Specifies the key identifier to use with the \fBntpdc\fR utility program, which uses a proprietary protocol specific to this implementation of \fBntpd\fR. The \fB\fIkey\fP\fR argument is a key identifier for the trusted key, where the value can be in the range 1 to 65,534, inclusive.
.RE
.sp .3
.ne 2
.mk
.na
\fBrevoke [\fIlogsec\fP]\fR
.ad
.sp .25
.RS 4n
Specifies the interval between re-randomization of certain cryptographic values used by the Autokey scheme, as a power of 2 in seconds. These values need to be updated frequently in order to deflect brute-force attacks on the algorithms; however, updating some values is a relatively expensive operation. The default interval is 16 (65,536 s or about 18 hours). For poll intervals above the specified interval, the values will be updated for every message sent.
.RE
.sp .3
.ne 2
.mk
.na
\fBtrustedkey \fIkey\fP [...]\fR
.ad
.sp .25
.RS 4n
Specifies the key identifiers which are trusted for the purposes of authenticating peers with symmetric key cryptography, as well as keys used by the \fBntpq\fR and \fBntpdc\fR programs. The authentication procedures require that both the local and remote servers share the same key and key identifier for this purpose, although different keys can be used with different servers. The \fB\fIkey\fP\fR arguments are 32-bit unsigned integers with values from 1 to 65,534.
.RE
.SS "Access Control Commands"
.sp .3
.ne 2
.mk
.na
\fBdiscard [ average \fIavg\fP ][ minimum \fImin\fP ] [ monitor \fIprob\fP ]\fR
.ad
.sp .25
.RS 4n
Set the parameters of the \fBlimited\fR facility which protects the server from client abuse. The \fBaverage\fR subcommand specifies the minimum average packet spacing, while the \fBminimum\fR subcommand specifies the minimum packet spacing. Packets that violate these minima are discarded and a kiss-o'-death packet returned if enabled. The default minimum average and minimum are 5 and 2, respectively. The monitor subcommand specifies the probability of discard for packets that overflow the rate-control window.
.RE
.sp .3
.ne 2
.mk
.na
\fBrestrict \fIaddress\fP [mask \fImask\fP] [\fIflag\fP][...]\fR
.ad
.sp .25
.RS 4n
The \fIaddress\fP argument expressed in dotted-quad form is the address of a host or network. Alternatively, the \fB\fIaddress\fP\fR argument can be a valid host DNS&nbsp;name. The \fImask\fP argument expressed in dotted-quad form defaults to \fB255.255.255.255\fR, meaning that the \fIaddress\fP is treated as the address of an individual host. A default entry (address \fB0.0.0.0\fR, mask \fB0.0.0.0\fR) is always included and is always the first entry in the list. Note that text string \fBdefault\fR, with no mask option, may be used to indicate the default entry.
.LP
In the current implementation, \fIflag\fP always restricts access, i.e., an entry with no flags indicates that free access to the server is to be given. The flags are not orthogonal, in that more restrictive flags will often make less restrictive ones redundant. The flags can generally be classed into two catagories, those which restrict time service and those which restrict informational queries and attempts to do run-time reconfiguration of the server. One or more of the following flags may be specified:
.RE
.sp .3
.ne 2
.mk
.na
\fBignore\fR
.ad
.sp .25
.RS 4n
Deny packets of all kinds, including \fBntpq\fR and \fBntpdc\fR queries.
.RE
.sp .3
.ne 2
.mk
.na
\fBkod\fR
.ad
.sp .25
.RS 4n
If this flag is set when an access violation occurs, a kiss-o'-death (KoD) packet is sent. KoD packets are rate limited to no more than one per second. If another KoD packet occurs within one second after the last one, the packet is dropped
.RE
.sp .3
.ne 2
.mk
.na
\fBlimited\fR
.ad
.sp .25
.RS 4n
Deny service if the packet spacing violates the lower limits specified in the \fBdiscard\fR command. A history of clients is kept using the monitoring capability of \fBntpd\fR. Thus, monitoring is always active as long as there is a restriction entry with the \fBlimited\fR flag.
.RE
.sp .3
.ne 2
.mk
.na
\fBlowpriotrap\fR
.ad
.sp .25
.RS 4n
Declare traps set by matching hosts to be low priority. The number of traps a server can maintain is limited (the current limit is 3). Traps are usually assigned on a first come, first served basis, with later trap requestors being denied service. This flag modifies the assignment algorithm by allowing low priority traps to be overridden by later requests for normal priority traps.
.RE
.sp .3
.ne 2
.mk
.na
\fBnomodify\fR
.ad
.sp .25
.RS 4n
Deny \fBntpq\fR and \fBntpdc\fR queries which attempt to modify the state of the server (i.e., run time reconfiguration). Queries which return information are permitted.
.RE
.sp .3
.ne 2
.mk
.na
\fBnoquery\fR
.ad
.sp .25
.RS 4n
Deny \fBntpq\fR and \fBntpdc\fR queries. Time service is not affected.
.RE
.sp .3
.ne 2
.mk
.na
\fBnopeer\fR
.ad
.sp .25
.RS 4n
Deny packets which would result in mobilizing a new association. &nbsp;This includes broadcast, symmetric-active and manycast client packets when a configured association does not exist.
.RE
.sp .3
.ne 2
.mk
.na
\fBnoserve\fR
.ad
.sp .25
.RS 4n
Deny all packets except \fBntpq\fR and \fBntpdc\fR queries.
.RE
.sp .3
.ne 2
.mk
.na
\fBnotrap\fR
.ad
.sp .25
.RS 4n
Decline to provide mode 6 control message trap service to matching hosts. The trap service is a subsystem of the \fBntpdq\fR control message protocol which is intended for use by remote event logging programs.
.RE
.sp .3
.ne 2
.mk
.na
\fBnotrust\fR
.ad
.sp .25
.RS 4n
Deny packets unless the packet is cryptographically authenticated.
.RE
.sp .3
.ne 2
.mk
.na
\fBntpport\fR
.ad
.sp .25
.RS 4n
This is actually a match algorithm modifier, rather than a restriction flag. Its presence causes the restriction entry to be matched only if the source port in the packet is the standard NTP UDP port (123). Both \fBntpport\fR and \fBnon-ntpport\fR may be specified. The \fBntpport\fR is considered more specific and is sorted later in the list.
.RE
.sp .3
.ne 2
.mk
.na
\fBversion\fR
.ad
.sp .25
.RS 4n
Deny packets that do not match the current NTP version.
.RE
.SS "Monitoring Commands"
.sp .3
.ne 2
.mk
.na
\fBstatistics \fIname\fP [...]\fR
.ad
.sp .25
.RS 4n
Enables writing of statistics records. Currently, six kinds of \fIname\fPstatistics are supported.
.RE
.sp .3
.ne 2
.mk
.na
\fBclockstats\fR
.ad
.sp .25
.RS 4n
Enables recording of clock driver statistics information. Each update received from a clock driver appends a line of the following form to the file generation set named \fBclockstats\fR:
.LP
\fB49213 525.624 127.127.4.1 93 226 00:08:29.606 D\fR
.LP
The first two fields show the date (Modified Julian Day) and time (seconds and fraction past UTC midnight). The next field shows the clock address in dotted-quad notation, The final field shows the last timecode received from the clock in decoded ASCII format, where meaningful. In some clock drivers a good deal of additional information can be gathered and displayed as well. See information specific to each clock for further details.
.RE
.sp .3
.ne 2
.mk
.na
\fBcryptostats\fR
.ad
.sp .25
.RS 4n
This option requires the OpenSSL cryptographic software library. It enables recording of cryptographic public key protocol information. Each message received by the protocol module appends a line of the following form to the file generation set named \fBcryptostats\fR:
.LP
\fB49213 525.624 127.127.4.1 \fImessage\fP\fR
.LP
The first two fields show the date (Modified Julian Day) and time (seconds and fraction past UTC midnight). The next field shows the peer address in dotted-quad notation, The final \fB\fImessage\fP\fR field includes the message type and certain ancillary information. See the Authentication Options page at file:///usr/share/doc/ntp/authopt.html for further information.
.RE
.sp .3
.ne 2
.mk
.na
\fBloopstats\fR
.ad
.sp .25
.RS 4n
Enables recording of loop filter statistics information. Each update of the local clock outputs a line of the following form to the file generation set named \fBloopstats\fR:
.LP
\fB50935 75440.031 0.000006019 13.778190 0.000351733 0.0133806	6\fR
.LP
The first two fields show the date (Modified Julian Day) and time (seconds and fraction past UTC midnight). The next five fields show time offset (seconds), frequency offset (parts per million - PPM), RMS jitter (seconds), Allan deviation (PPM) and clock discipline time constant.
.RE
.sp .3
.ne 2
.mk
.na
\fBpeerstats\fR
.ad
.sp .25
.RS 4n
Enables recording of peer statistics information. This includes statistics records of all peers of a NTP server and of special signals, where present and configured. Each valid update appends a line of the following form to the current element of a file generation set named \fBpeerstats\fR:
.LP
\fB48773 10847.650 127.127.4.1 9714 -0.001605376 0.000000000 0.001424877 0.000958674\fR
.LP
The first two fields show the date (Modified Julian Day) and time (seconds and fraction past UTC midnight). The next two fields show the peer address in dotted-quad notation and status, respectively. The status field is encoded in hex in the format described in Appendix B of the NTP specification RFC 1305. The final four fields show the offset, delay, dispersion and RMS jitter, all in seconds.
.RE
.sp .3
.ne 2
.mk
.na
\fBrawstats\fR
.ad
.sp .25
.RS 4n
Enables recording of raw-timestamp statistics information. This includes statistics records of all peers of a NTP server and of special signals, where present and configured. Each NTP message received from a peer or clock driver appends a line of the following form to the file generation set named \fBrawstats\fR:
.LP
\fB50928 2132.2543 128.4.1.1 128.4.1.20 3102453281.2584327000 3102453281.258622800031 02453332.2540806000 3102453332.2541458000\fR
.LP
The first two fields show the date (Modified Julian Day) and time (seconds and fraction past UTC midnight). The next two fields show the remote peer or clock address followed by the local address in dotted-quad notation, The final four fields show the originate, receive, transmit and final NTP timestamps in order. The timestamp values are as received and before processing by the various data smoothing and mitigation algorithms.
.RE
.sp .3
.ne 2
.mk
.na
\fBsysstats\fR
.ad
.sp .25
.RS 4n
Enables recording of \fBntpd\fR statistics counters on a periodic basis. Each hour a line of the following form is appended to the file generation set named \fBsysstats\fR:
.LP
\fB50928 2132.2543 36000 81965 0 9546 56 71793 512 540 10 147\fR
.LP
The first two fields show the date (Modified Julian Day) and time (seconds and fraction past UTC midnight). The remaining ten fields show the statistics counter values accumulated since the last generated line.
.sp .3
.ne 2
.mk
.na
Time since restart \fB36000\fR:
Time in hours since the system was last rebooted.
.sp .3
.ne 2
.mk
.na
Packets received \fB81965\fR:
Total number of packets received.
.sp .3
.ne 2
.mk
.na
Packets processed \fB0\fR:
Number of packets received in response to previous packets sent
.sp .3
.ne 2
.mk
.na
Current version \fB9546\fR:
Number of packets matching the current NTP version.
.sp .3
.ne 2
.mk
.na
Previous version \fB56\fR:
Number of packets matching the previous NTP version.
.sp .3
.ne 2
.mk
.na
Bad version \fB71793\fR:
Number of packets matching neither NTP version.
.sp .3
.ne 2
.mk
.na
Access denied \fB512\fR:
Number of packets denied access for any reason.
.sp .3
.ne 2
.mk
.na
Bad length or format \fB540\fR:
Number of packets with invalid length, format or port number.
.sp .3
.ne 2
.mk
.na
Bad authentication \fB10\fR:
Number of packets not verified as authentic.
.sp .3
.ne 2
.mk
.na
Rate exceeded \fB147\fR:
Number of packets discarded due to rate limitation.
.RE
.sp .3
.ne 2
.mk
.na
\fBstatsdir \fIdirectory_path\fP\fR
.ad
.sp .25
.RS 4n
Indicates the full path of a directory where statistics files should be created (see below). This keyword allows the (otherwise constant) \fBfilegen\fR filename prefix to be modified for file generation sets, which is useful for handling statistics logs.
.RE
.sp .3
.ne 2
.mk
.na
\fBfilegen \fIname\fP [file \fIfilename\fP] [type \fItypename\fP] [link | nolink] [enable | disable]\fR
.ad
.sp .25
.RS 4n

components/ntp/Solaris/ntp.sh
#!/sbin/sh
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#

#
# Copyright (c) 2009, 2011, Oracle and/or its affiliates. All rights reserved.
#

# Standard prolog
#
. /lib/svc/share/smf_include.sh

if [ -z $SMF_FMRI ]; then
        echo "SMF framework variables are not initialized."
        exit $SMF_EXIT_ERR
fi

#
# Is NTP configured?
#
if [ ! -f /etc/inet/ntp.conf ]; then
	echo "Error: Configuration file '/etc/inet/ntp.conf' not found." \
	    "  See ntpd(1M)."
	exit $SMF_EXIT_ERR_CONFIG
fi

# Disable globbing to prevent privilege escalations by users authorized
# to set property values for the NTP service.
set -f 

#
# Build the command line flags
#
shift $#
set -- -p /var/run/ntp.pid
# We allow a step larger than the panic value of 17 minutes only 
# once when ntpd starts up. If always_all_large_step is true, 
# then we allow this each time ntpd starts. Otherwise, we allow
# it only the very first time ntpd starts after a boot. We 
# check that by making ntpd write its pid to a file in /var/run.

val=`svcprop -c -p config/always_allow_large_step $SMF_FMRI`
if [ "$val" = "true" ] || \
    [ ! -f /var/run/ntp.pid ]; then
        set -- "$@" -g
fi

# Auth was off by default in xntpd now the default is on. Better have a way
# to turn it off again. Also check for the obsolete "authentication" keyword.
val=`svcprop -c -p config/no_auth_required $SMF_FMRI`
if [ ! "$val" = "true" ]; then
        val=`/usr/bin/nawk '/^[ \t]*#/{next}
            /^[ \t]*authentication[ \t]+no/ {
                printf("true", $2)
                next } ' /etc/inet/ntp.conf`
fi
[ "$val" = "true" ] && set -- "$@" --authnoreq

# Set up logging if requested.
logfile=`svcprop -c -p config/logfile $SMF_FMRI`
val=`svcprop -c -p config/verbose_logging $SMF_FMRI`
[ "$val" = "true" ] && [ -n "$logfile" ]  && set -- "$@" -l $logfile

# Register with mDNS.
val=`svcprop -c -p config/mdnsregister $SMF_FMRI`
mdns=`svcprop -c -p general/enabled svc:/network/dns/multicast:default`
[ "$val" = "true" ] && [ "$mdns" = "true" ] && set -- "$@" -m

# We used to support the slewalways keyword, but that was a Sun thing
# and not in V4. Look for "slewalways yes" and set the new slew option.
val=`svcprop -c -p config/slew_always $SMF_FMRI`
if [ ! "$val" = "true" ]; then
	val=`/usr/bin/nawk '/^[ \t]*#/{next}
	    /^[ \t]*slewalways[ \t]+yes/ {
        	printf("true", $2)
        	next } ' /etc/inet/ntp.conf`
fi
[ "$val" = "true" ] && set -- "$@" --slew

# Set up debugging.
deb=`svcprop -c -p config/debuglevel $SMF_FMRI`

# Start the daemon. If debugging is requested, put it in the background, 
# since it won't do it on it's own.
if [ "$deb" -gt 0 ]; then
	/usr/lib/inet/ntpd "$@" --set-debug-level=$deb >/var/ntp/ntp.debug &
else
	/usr/lib/inet/ntpd "$@"
fi

# Now, wait for the first sync, if requested.
val=`svcprop -c -p config/wait_for_sync $SMF_FMRI`
[ "$val" = "true" ] && /usr/lib/inet/ntp-wait

exit $SMF_EXIT_OK
